lib/rgl/bidirectional_adjacency.rb

```diff
@@ -0,0 +1,87 @@
+# bidirectional_adjacency.rb
+#
+require 'rgl/adjacency'
+require 'rgl/bidirectional'
+
+module RGL
+
+  # BGL defines the concept BidirectionalGraph as follows:
+  #
+  # The BidirectionalGraph concept refines IncidenceGraph and adds the
+  # requirement for efficient access to the in-edges of each vertex. This
+  # concept is separated from IncidenceGraph because, for directed graphs,
+  # efficient access to in-edges typically requires more storage space,
+  # and many algorithms do not require access to in-edges. For undirected
+  # graphs, this is not an issue; because the in_edges() and out_edges()
+  # functions are the same, they both return the edges incident to the vertex.
+
+  # This implementation simply creates an internal DirectedAdjacencyGraph to store
+  # the in-edges and overrides methods to ensure that the out and in graphs
+  # remain synchronized.
+  #
+  class BidirectionalAdjacencyGraph < DirectedAdjacencyGraph
+
+    include BidirectionalGraph
+
+    # @see DirectedAdjacencyGraph#initialize
+    def initialize(edgelist_class = Set, *other_graphs)
+      super(edgelist_class, *other_graphs)
+      @reverse = self.reverse
+    end
+
+    # We don't need to override add_vertex() because the reverse graph doesn't need to
+    # contain any unconnected vertices. Vertices will be added by add_edge() as
+    # required.
+
+    # @see MutableGraph#add_edge.
+    def add_edge(u, v)
+      super(u, v)
+      @reverse.add_edge(v, u)
+    end
+
+    # @see MutableGraph#remove_vertex.
+    def remove_vertex(v)
+      super(v)
+      @reverse.remove_vertex(v)
+    end
+
+    # @see MutableGraph::remove_edge.
+    def remove_edge(u, v)
+      super(u, v)
+      @reverse.remove_edge(v, u)
+    end
+
+    # @see Graph#has_edge?
```

Comment:
These two should be defined in module `BidirectionalGraph`:

```ruby
    def has_in_edge?(u, v)
      raise NotImplementedError
    end

    alias :has_out_edge? :has_edge?
```

Reply:
`has_edge?` is apparently not defined in any module (only in `adjacency.rb`). Should it be added to `base.rb`?

Reply:
Oh, yes! We could provide a default implementation in `base.rb`:

```ruby
   def has_edge?(u, v)
      each_adjacent(u) do |w|
        return true if v == w
      end
      return false
    end
```

Reply:
Done.